akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala

```diff
@@ -33,34 +33,24 @@ private[akka] object InputStreamSinkStage {
 /**
  * INTERNAL API
  */
-private[akka] class InputStreamSinkStage(timeout: FiniteDuration) extends SinkStage[ByteString, InputStream]("InputStreamSinkStage") {
+private[akka] class InputStreamSinkStage(timeout: FiniteDuration) extends SinkStage[ByteString, InputStream]("InputStreamSink") {
   val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
   require(maxBuffer > 0, "Buffer size must be greater than 0")
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, InputStream) = {
-
     val dataQueue = new LinkedBlockingDeque[StreamToAdapterMessage](maxBuffer + 1)
-    var pullRequestIsSent = true
 
     val logic = new GraphStageLogic(shape) with StageWithCallback {
+      var pullRequestIsSent = true
 
       private val callback: AsyncCallback[AdapterToStageMessage] =
-        getAsyncCallback(onAsyncMessage)
-
-      override def wakeUp(msg: AdapterToStageMessage): Unit = {
-        if (!isClosed(in)) {
-          Future(callback.invoke(msg))(interpreter.materializer.executionContext)
-        }
-      }
-
-      private def onAsyncMessage(event: AdapterToStageMessage): Unit =
-        event match {
-          case ReadElementAcknowledgement ⇒
-            sendPullIfAllowed()
-          case Close ⇒
-            completeStage()
+        getAsyncCallback {
+          case ReadElementAcknowledgement ⇒ sendPullIfAllowed()
+          case Close                      ⇒ completeStage()
         }
 
+      override def wakeUp(msg: AdapterToStageMessage): Unit = if (!isClosed(in)) callback.invoke(msg)
```

Comment:
this is wrong because it evaluates `isClosed(in)` from a different execution context, and it is also not needed: simply pass `callback.invoke _` to the handler’s constructor